### PR TITLE
APPROVED : IP Token end-to-end encryption for provider key transit security

### PR DESCRIPTION
**Pull request by alchezar**
# Summary

- Move `IpTokenCrypto` from `iron_control_api` to shared `iron_secrets` crate so both server and client can use it
- Encrypt provider key as IP Token in `/api/v1/agents/provider-key` response (Flow 2 was returning plaintext)
- Decrypt IP Token in runtime clients (`KeyFetcher`, `BudgetClient`) instead of using encrypted string as-is (Flow 1 was sending `AES256:…` to OpenAI)
- Fix silent plaintext fallback: `None` arm in both clients now returns `Err` instead of passing ciphertext to the LLM provider
- Add startup validation: `ProxyConfig` and `BudgetClient` fail loudly at construction when `IP_TOKEN_KEY` is absent
- Wire `IP_TOKEN_KEY` env var in `LlmRouter` for production decryption
- Eliminate `ProviderKey` duplication between `key_fetcher` and `budget_client`
- Delete orphaned `iron_control_api/src/ip_token.rs` (6 693 bytes, never compiled)

## Problem

`IP_TOKEN_KEY` encryption was implemented on the server but not working end-to-end. Three separate issues broke the provider key transit security layer:

- **Flow 1 (Budget Handshake)** — client used encrypted `AES256:…` string as API key without decrypting → 401 from OpenAI/Anthropic, ciphertext leaked to provider
- **Flow 2 (Feature 014, provider-key)** — server returned plaintext provider key, skipping IP Token encryption entirely → working but unprotected
- **Silent fallback (both clients)** — `None` arm passed ciphertext verbatim when `IP_TOKEN_KEY` env var was absent; misconfiguration was invisible until the first LLM call failed with 401

## Security invariant restored

Plaintext provider keys never appear in HTTP response bodies — only `AES256:{iv}:{ciphertext}:{tag}` format is transmitted. Misconfigured runtimes fail loudly at startup, not silently at the first LLM request.

## Test plan

- [x] 16 unit tests: roundtrip, empty/long strings, format validation, tamper detection, wrong key, key length
- [x] 4 e2e tests: both flows encrypt→decrypt, corrupted token handling, cross-flow nonce uniqueness
- [x] 2 integration tests via real `KeyFetcher` / `BudgetClient` API with mock HTTP server (issue-001, issue-002 regression guards with 5-section format)
- [x] 1 startup validation test: `run_proxy_fails_loudly_when_ip_token_key_absent`
- [x] `cargo build --all-features` and `cargo clippy -- -D warnings` pass

### CODE REVIEW

**Review by wanguardd — DISMISSED** [review]
## Review Summary

**Outcome**: NOT READY

### Verification Results

| Check | Status |
|-------|--------|
| Build | Not run (working tree has unstaged changes) |
| Code analysis | Complete — 9-round review |
| Prior reviews | None |

---

## Code Review (9 Rounds)

| Round | Category | Status |
|-------|----------|--------|
| 0 | ORGANIZATIONAL | FINDINGS |
| 1 | SCOPE | PASS |
| 2 | CORRECTNESS | FINDINGS (Critical) |
| 3 | SECURITY | FINDINGS (Critical) |
| 4 | REQUIREMENTS | PASS |
| 5 | DESIGN | FINDINGS |
| 6 | QUALITY | FINDINGS |
| 7 | TESTING | FINDINGS |
| 8 | DOCUMENTATION | FINDINGS |

---

## Findings

### Finding 1 (CORRECTNESS / SECURITY) — Silent Plaintext Fallback in `key_fetcher.rs`

**Severity**: Critical — re-introduces Bug 1 under missing env var

`module/iron_runtime/src/llm_router/key_fetcher.rs`, `fetch_from_server()`:

```rust
let api_key = match &self.ip_token_crypto {
    Some(crypto) => crypto
        .decrypt(&data.ip_token)
        ...
        .to_string(),
    None => data.ip_token,   // ← line 202
};
```

When `IP_TOKEN_KEY` is absent from the environment, `ip_token_crypto` is `None`. The `None` arm takes `data.ip_token` verbatim — an opaque `AES256:{IV}:{ciphertext}:{tag}` string — and uses it as the provider API key. Every LLM call fails with 401 from OpenAI/Anthropic, and the encrypted string is transmitted to a third-party API endpoint.

**Required fix**: Return an error when decryption is required but no key is available. The correct pattern:

```rust
let api_key = match &self.ip_token_crypto {
    Some(crypto) => crypto
        .decrypt(&data.ip_token)
        ...
        .to_string(),
    None => return Err(LlmRouterError::KeyFetch(
        "IP_TOKEN_KEY not configured — cannot decrypt IP Token".into()
    )),
};
```

---

### Finding 2 (CORRECTNESS / SECURITY) — Silent Plaintext Fallback in `budget_client.rs`

**Severity**: Critical — same bug independently in BudgetClient

`module/iron_cost/src/budget_client.rs`, handshake path:

```rust
let api_key = match &self.ip_token_crypto {
    Some(crypto) => crypto
        .decrypt(&data.ip_token)
        .map_err(|e| BudgetClientError::IpTokenDecrypt(e.to_string()))?
        .to_string(),
    None => data.ip_token,   // ← line 281: ciphertext used as API key
};
```

Same problem as Finding 1, in the budget handshake flow (Protocol 005). When `IP_TOKEN_KEY` is absent, the handshake gives the agent a `AES256:...` ciphertext as its API key.

**Required fix**: Fail loudly when `ip_token_key` is absent but the server returned an encrypted token. A deployment without `IP_TOKEN_KEY` must fail at startup, not silently at runtime.

---

### Finding 3 (SECURITY) — `.to_string()` Destroys `Zeroizing<String>` Memory Protection

**Severity**: High

`IpTokenCrypto::decrypt()` returns `Zeroizing<String>`, which zeroes the heap allocation on drop. Both client sites immediately convert this to a plain `String`:

`key_fetcher.rs:201`:
```rust
.decrypt(&data.ip_token)
.map_err(...)?
.to_string(),   // ← creates unprotected String copy; Zeroizing drops immediately after
```

`budget_client.rs:279-280`:
```rust
.decrypt(&data.ip_token)
.map_err(...)?
.to_string(),   // ← same issue
```

`Zeroizing<String>` is a `Deref<Target=String>` and `AsRef<str>`. Calling `.to_string()` creates a fresh, unprotected heap allocation — the decrypted key persists in memory as a plain `String` indefinitely. The entire point of `Zeroizing` is defeated at the call site.

**Required fix**: Pass `Zeroizing<String>` through to where the key is actually used, or at minimum store it in a `Zeroizing`-wrapped field. Do not call `.to_string()` on a `Zeroizing<String>`.

---

### Finding 4 (ORGANIZATIONAL) — Orphaned `iron_control_api/src/ip_token.rs`

**Severity**: High

`module/iron_control_api/src/ip_token.rs` contains a full 222-line implementation of `IpTokenCrypto` and `IpTokenError`. In `lib.rs`, there is no `mod ip_token;` declaration. Instead:

```rust
// lib.rs:211
pub use iron_secrets::ip_token;
```

This re-exports the `iron_secrets` module as `iron_control_api::ip_token`. The local `src/ip_token.rs` file is never compiled — it is dead code. It's also a duplicate of the `iron_secrets` implementation, violating the Anti-Duplication Principle.

**Evidence**:
```bash
$ git show HEAD:module/iron_control_api/src/lib.rs | grep "ip_token"
pub use iron_secrets::ip_token;
# No "mod ip_token;" present
```

**Required fix**: Delete `module/iron_control_api/src/ip_token.rs`. The re-export in `lib.rs` already provides `iron_control_api::ip_token` for consumers that import it that way.

---

### Finding 5 (QUALITY) — Codestyle Violations in `iron_secrets/src/ip_token.rs`

**Severity**: High — the primary new file uses `cargo fmt` style throughout

The project codestyle mandates: new-line braces, spaces inside `#[ derive( ) ]`, spaces inside `< >` and function parameter parentheses. `iron_secrets/src/ip_token.rs` violates all three:

```rust
// Actual (cargo fmt style):
pub struct IpTokenCrypto {
    cipher: Aes256Gcm,
}

impl Debug for IpTokenCrypto {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum IpTokenError {
```

Required project style:
```rust
pub struct IpTokenCrypto
{
  cipher : Aes256Gcm,
}

impl Debug for IpTokenCrypto
{
  fn fmt( &self, f : &mut Formatter< '_ > ) -> FmtResult
  {

#[ derive( Debug, Clone, PartialEq, Eq ) ]
pub enum IpTokenError
{
```

The entire 222-line file needs reformatting. The 4-space indent vs project 2-space indent is also present throughout.

---

### Finding 6 (QUALITY / DOCUMENTATION) — Missing Bug Fix Documentation

**Severity**: High

Two bugs were fixed in this PR (server returned plaintext; client used ciphertext as API key). The project mandates specific fix documentation for every bug fix:

1. **Test**: `bug_reproducer(issue-NNN)` marker + 5-section doc comment (Root Cause, Why Not Caught, Fix Applied, Prevention, Pitfall)
2. **Source**: 3-field comment at the fix site (`Fix(issue-NNN)`, `Root cause:`, `Pitfall:`)

```bash
$ grep -rn "bug_reproducer" module/iron_secrets/ module/iron_runtime/ module/iron_cost/ module/iron_control_api/tests/ip_token_e2e_test.rs
# (no output — zero occurrences)
```

Neither `ip_token_test.rs` nor `ip_token_e2e_test.rs` contains `bug_reproducer` markers or fix documentation. Neither fixed source file (`key_fetcher.rs`, `budget_client.rs`) contains 3-field fix comments.

**Required**: Create at minimum one `bug_reproducer`-marked test per bug with the 5-section format, and add 3-field source comments at each fix site.

---

### Finding 7 (TESTING) — `KeyFetcher` Decryption Path Has Zero Integration Tests

**Severity**: High

The e2e tests in `ip_token_e2e_test.rs` verify that:
1. The server encrypts the provider key ✓
2. `IpTokenCrypto::decrypt()` can decrypt it ✓

But they do NOT test that `KeyFetcher::fetch_from_server()` correctly decrypts the server response. The test calls `IpTokenCrypto::decrypt()` directly:

```rust
// ip_token_e2e_test.rs — client side of "e2e" test:
let client_crypto = IpTokenCrypto::new(&TEST_IP_TOKEN_KEY).expect("...");
let decrypted = client_crypto.decrypt(ip_token).expect("...");
```

`KeyFetcher` is never instantiated in any test. The actual production code path — `KeyFetcher::new(..., Some(ip_token_key))` → `fetch_from_server()` → `ip_token_crypto.decrypt()` — has zero test coverage. Finding 1 (the `None` fallback arm) could only be caught by such a test.

**Required**: Add an integration test that creates a `KeyFetcher` with `ip_token_key: Some([...])`, points it at a test server that returns an encrypted `ip_token` response, and verifies `get_key()` returns the correct plaintext key.

---

### Finding 8 (ORGANIZATIONAL) — File Creation Protocol: 3 New Files, 0 readme.md Updates

**Severity**: High

Three new permanent files were added without updating the corresponding `readme.md` Responsibility Tables:

| New File | Directory `readme.md` | Status |
|----------|-----------------------|--------|
| `iron_secrets/src/ip_token.rs` | `iron_secrets/src/readme.md` | Not listed |
| `iron_secrets/tests/ip_token_test.rs` | `iron_secrets/tests/readme.md` | Not listed |
| `iron_control_api/tests/ip_token_e2e_test.rs` | `iron_control_api/tests/readme.md` | Not listed |

Evidence for `iron_secrets/tests/readme.md` (only lists 2 files, both pre-existing):
```
| `crypto_test.rs` | Test cryptographic operations and key derivation |
| `smoke_test.rs` | Test basic functionality and integration smoke tests |
```

`ip_token_test.rs` is absent.

**Required**: Add one row per new file to the appropriate `readme.md` Responsibility Table, following the mandatory File Creation Protocol (5-step workflow).

---

### Finding 9 (ORGANIZATIONAL) — Incorrect `Entity` Header in Responsibility Table

**Severity**: Medium

`module/iron_control_api/tests/readme.md` uses `| Entity |` as the first column header instead of the required `| File |`:

```
| Entity | Responsibility | Input→Output | Out of Scope |
```

Per `organizational_principles.rulebook.md` § Table Quality Standard 4: "Use 'File' for directories, 'Component' for architecture". This is a directory listing, so the header must be `| File |`.

**Required fix**: Replace `Entity` with `File` in the table header.

---

### Finding 10 (DESIGN) — Fallback-by-Default Pattern for a Security-Critical Configuration

**Severity**: Medium

The `ip_token_key: Option<[u8; 32]>` pattern in both `KeyFetcher::new()` and `BudgetClientConfig` allows the system to silently operate without encryption. This is a fail-open design for a security-critical feature.

The `IP_TOKEN_KEY` env var should either:
1. Be required at startup (fail-fast if absent when using server mode), or
2. If optional, the `None` arm must produce an error, not a silent passthrough

The current design requires developers to remember to set `IP_TOKEN_KEY` — a configuration gap that produces silent data corruption (ciphertext sent to OpenAI) rather than a startup failure.

---

### Finding 11 (TESTING) — E2E Tests Call Crypto Directly, Not Through Client APIs

**Severity**: Medium

`e2e_handshake_server_encrypts_client_decrypts` and similar tests verify the crypto algebra but not the system integration. The "client side" of each test creates a bare `IpTokenCrypto` instance and calls `decrypt()` directly, bypassing:
- `BudgetClient::handshake()`
- `KeyFetcher::get_key()` and `fetch_from_server()`
- The `ip_token_key` → `IpTokenCrypto` initialization path

True e2e tests would go through the actual client APIs (`BudgetClient::handshake()`, `KeyFetcher::get_key()`) and verify `get_provider_key()` returns the correct plaintext — proving the production path works. These tests are integration tests for the crypto module, not e2e tests for the full flow.

---

## Positive Observations

1. **Correct architectural decision**: Moving `IpTokenCrypto` to `iron_secrets` so both server and client can use the same implementation is the right call. The shared crypto is clean and well-structured.

2. **`iron_secrets/src/ip_token.rs` logic is sound**: The AES-256-GCM implementation (nonce generation, ciphertext format `AES256:{IV}:{ciphertext}:{tag}`, tag splitting) is correct. The `IpTokenError` variants cover all failure modes explicitly.

3. **Server-side encryption wired correctly**: Both `handshake.rs` and `agent_provider_key.rs` correctly call `state.ip_token_crypto.encrypt(&provider_key)` before returning the response. The security invariant "plaintext key never in HTTP response" holds for the server side.

4. **`IP_TOKEN_KEY` env var wired in router**: `router.rs` correctly reads `IP_TOKEN_KEY` via `parse_ip_token_key_from_env()` and passes it to both `KeyFetcher::new()` and `ProxyConfig`. The env var parsing handles missing/malformed values with appropriate tracing warnings.

5. **16 unit tests for crypto core**: `ip_token_test.rs` covers roundtrip, empty/long strings, format validation, tamper detection, wrong key, and key length validation. Good coverage of the crypto primitive.

6. **`GetProviderKeyResponse` field correctly renamed to `ip_token`**: The response struct field name matches the encrypted nature of the data — `provider_key` implied plaintext, `ip_token` correctly signals it's an encrypted token.

---

## Required Changes Before Merge

1. **Fix CRIT**: Replace `None => data.ip_token` with `None => return Err(...)` in both `key_fetcher.rs:202` and `budget_client.rs:281`
2. **Fix CRIT**: Propagate `Zeroizing<String>` without calling `.to_string()` at `key_fetcher.rs:201` and `budget_client.rs:280`
3. **Fix HIGH**: Delete `module/iron_control_api/src/ip_token.rs` (dead orphaned file)
4. **Fix HIGH**: Reformat `iron_secrets/src/ip_token.rs` to project codestyle (new-line braces, spaces in `#[ derive( ) ]` and `< >`, 2-space indent)
5. **Fix HIGH**: Add `bug_reproducer`-marked tests with 5-section docs for both fixed bugs; add 3-field fix comments in source
6. **Fix HIGH**: Add integration test for `KeyFetcher` decryption path (not direct crypto calls)
7. **Fix HIGH**: Update 3 `readme.md` files with new file entries
8. **Fix MED**: Change `Entity` → `File` in `iron_control_api/tests/readme.md` table header

**Review by wanguardd — DISMISSED** [review]
## Review Summary

**Outcome**: NOT READY

### Verification Results

| Check | Status |
|-------|--------|
| Build | Not run (working tree has unstaged changes) |
| Code analysis | Complete — 9-round review |
| Prior reviews | None |

---

## Code Review (9 Rounds)

| Round | Category | Status |
|-------|----------|--------|
| 0 | ORGANIZATIONAL | FINDINGS |
| 1 | SCOPE | PASS |
| 2 | CORRECTNESS | FINDINGS (Critical) |
| 3 | SECURITY | FINDINGS (Critical) |
| 4 | REQUIREMENTS | PASS |
| 5 | DESIGN | FINDINGS |
| 6 | QUALITY | FINDINGS |
| 7 | TESTING | FINDINGS |
| 8 | DOCUMENTATION | FINDINGS |

---

## Findings

### Finding 1 (CORRECTNESS / SECURITY) — Silent Plaintext Fallback in `key_fetcher.rs`

**Severity**: Critical — re-introduces Bug 1 under missing env var

`module/iron_runtime/src/llm_router/key_fetcher.rs`, `fetch_from_server()`:

```rust
let api_key = match &self.ip_token_crypto {
    Some(crypto) => crypto
        .decrypt(&data.ip_token)
        ...
        .to_string(),
    None => data.ip_token,   // ← line 202
};
```

When `IP_TOKEN_KEY` is absent from the environment, `ip_token_crypto` is `None`. The `None` arm takes `data.ip_token` verbatim — an opaque `AES256:{IV}:{ciphertext}:{tag}` string — and uses it as the provider API key. Every LLM call fails with 401 from OpenAI/Anthropic, and the encrypted string is transmitted to a third-party API endpoint.

**Required fix**: Return an error when decryption is required but no key is available. The correct pattern:

```rust
let api_key = match &self.ip_token_crypto {
    Some(crypto) => crypto
        .decrypt(&data.ip_token)
        ...
        .to_string(),
    None => return Err(LlmRouterError::KeyFetch(
        "IP_TOKEN_KEY not configured — cannot decrypt IP Token".into()
    )),
};
```

---

### Finding 2 (CORRECTNESS / SECURITY) — Silent Plaintext Fallback in `budget_client.rs`

**Severity**: Critical — same bug independently in BudgetClient

`module/iron_cost/src/budget_client.rs`, handshake path:

```rust
let api_key = match &self.ip_token_crypto {
    Some(crypto) => crypto
        .decrypt(&data.ip_token)
        .map_err(|e| BudgetClientError::IpTokenDecrypt(e.to_string()))?
        .to_string(),
    None => data.ip_token,   // ← line 281: ciphertext used as API key
};
```

Same problem as Finding 1, in the budget handshake flow (Protocol 005). When `IP_TOKEN_KEY` is absent, the handshake gives the agent a `AES256:...` ciphertext as its API key.

**Required fix**: Fail loudly when `ip_token_key` is absent but the server returned an encrypted token. A deployment without `IP_TOKEN_KEY` must fail at startup, not silently at runtime.

---

### Finding 3 (SECURITY) — `.to_string()` Destroys `Zeroizing<String>` Memory Protection

**Severity**: High

`IpTokenCrypto::decrypt()` returns `Zeroizing<String>`, which zeroes the heap allocation on drop. Both client sites immediately convert this to a plain `String`:

`key_fetcher.rs:201`:
```rust
.decrypt(&data.ip_token)
.map_err(...)?
.to_string(),   // ← creates unprotected String copy; Zeroizing drops immediately after
```

`budget_client.rs:279-280`:
```rust
.decrypt(&data.ip_token)
.map_err(...)?
.to_string(),   // ← same issue
```

`Zeroizing<String>` is a `Deref<Target=String>` and `AsRef<str>`. Calling `.to_string()` creates a fresh, unprotected heap allocation — the decrypted key persists in memory as a plain `String` indefinitely. The entire point of `Zeroizing` is defeated at the call site.

**Required fix**: Pass `Zeroizing<String>` through to where the key is actually used, or at minimum store it in a `Zeroizing`-wrapped field. Do not call `.to_string()` on a `Zeroizing<String>`.

---

### Finding 4 (ORGANIZATIONAL) — Orphaned `iron_control_api/src/ip_token.rs`

**Severity**: High

`module/iron_control_api/src/ip_token.rs` contains a full 222-line implementation of `IpTokenCrypto` and `IpTokenError`. In `lib.rs`, there is no `mod ip_token;` declaration. Instead:

```rust
// lib.rs:211
pub use iron_secrets::ip_token;
```

This re-exports the `iron_secrets` module as `iron_control_api::ip_token`. The local `src/ip_token.rs` file is never compiled — it is dead code. It's also a duplicate of the `iron_secrets` implementation, violating the Anti-Duplication Principle.

**Evidence**:
```bash
$ git show HEAD:module/iron_control_api/src/lib.rs | grep "ip_token"
pub use iron_secrets::ip_token;
# No "mod ip_token;" present
```

**Required fix**: Delete `module/iron_control_api/src/ip_token.rs`. The re-export in `lib.rs` already provides `iron_control_api::ip_token` for consumers that import it that way.

---

### Finding 5 (QUALITY) — Codestyle Violations in `iron_secrets/src/ip_token.rs`

**Severity**: High — the primary new file uses `cargo fmt` style throughout

The project codestyle mandates: new-line braces, spaces inside `#[ derive( ) ]`, spaces inside `< >` and function parameter parentheses. `iron_secrets/src/ip_token.rs` violates all three:

```rust
// Actual (cargo fmt style):
pub struct IpTokenCrypto {
    cipher: Aes256Gcm,
}

impl Debug for IpTokenCrypto {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum IpTokenError {
```

Required project style:
```rust
pub struct IpTokenCrypto
{
  cipher : Aes256Gcm,
}

impl Debug for IpTokenCrypto
{
  fn fmt( &self, f : &mut Formatter< '_ > ) -> FmtResult
  {

#[ derive( Debug, Clone, PartialEq, Eq ) ]
pub enum IpTokenError
{
```

The entire 222-line file needs reformatting. The 4-space indent vs project 2-space indent is also present throughout.

---

### Finding 6 (QUALITY / DOCUMENTATION) — Missing Bug Fix Documentation

**Severity**: High

Two bugs were fixed in this PR (server returned plaintext; client used ciphertext as API key). The project mandates specific fix documentation for every bug fix:

1. **Test**: `bug_reproducer(issue-NNN)` marker + 5-section doc comment (Root Cause, Why Not Caught, Fix Applied, Prevention, Pitfall)
2. **Source**: 3-field comment at the fix site (`Fix(issue-NNN)`, `Root cause:`, `Pitfall:`)

```bash
$ grep -rn "bug_reproducer" module/iron_secrets/ module/iron_runtime/ module/iron_cost/ module/iron_control_api/tests/ip_token_e2e_test.rs
# (no output — zero occurrences)
```

Neither `ip_token_test.rs` nor `ip_token_e2e_test.rs` contains `bug_reproducer` markers or fix documentation. Neither fixed source file (`key_fetcher.rs`, `budget_client.rs`) contains 3-field fix comments.

**Required**: Create at minimum one `bug_reproducer`-marked test per bug with the 5-section format, and add 3-field source comments at each fix site.

---

### Finding 7 (TESTING) — `KeyFetcher` Decryption Path Has Zero Integration Tests

**Severity**: High

The e2e tests in `ip_token_e2e_test.rs` verify that:
1. The server encrypts the provider key ✓
2. `IpTokenCrypto::decrypt()` can decrypt it ✓

But they do NOT test that `KeyFetcher::fetch_from_server()` correctly decrypts the server response. The test calls `IpTokenCrypto::decrypt()` directly:

```rust
// ip_token_e2e_test.rs — client side of "e2e" test:
let client_crypto = IpTokenCrypto::new(&TEST_IP_TOKEN_KEY).expect("...");
let decrypted = client_crypto.decrypt(ip_token).expect("...");
```

`KeyFetcher` is never instantiated in any test. The actual production code path — `KeyFetcher::new(..., Some(ip_token_key))` → `fetch_from_server()` → `ip_token_crypto.decrypt()` — has zero test coverage. Finding 1 (the `None` fallback arm) could only be caught by such a test.

**Required**: Add an integration test that creates a `KeyFetcher` with `ip_token_key: Some([...])`, points it at a test server that returns an encrypted `ip_token` response, and verifies `get_key()` returns the correct plaintext key.

---

### Finding 8 (ORGANIZATIONAL) — File Creation Protocol: 3 New Files, 0 readme.md Updates

**Severity**: High

Three new permanent files were added without updating the corresponding `readme.md` Responsibility Tables:

| New File | Directory `readme.md` | Status |
|----------|-----------------------|--------|
| `iron_secrets/src/ip_token.rs` | `iron_secrets/src/readme.md` | Not listed |
| `iron_secrets/tests/ip_token_test.rs` | `iron_secrets/tests/readme.md` | Not listed |
| `iron_control_api/tests/ip_token_e2e_test.rs` | `iron_control_api/tests/readme.md` | Not listed |

Evidence for `iron_secrets/tests/readme.md` (only lists 2 files, both pre-existing):
```
| `crypto_test.rs` | Test cryptographic operations and key derivation |
| `smoke_test.rs` | Test basic functionality and integration smoke tests |
```

`ip_token_test.rs` is absent.

**Required**: Add one row per new file to the appropriate `readme.md` Responsibility Table, following the mandatory File Creation Protocol (5-step workflow).

---

### Finding 9 (ORGANIZATIONAL) — Incorrect `Entity` Header in Responsibility Table

**Severity**: Medium

`module/iron_control_api/tests/readme.md` uses `| Entity |` as the first column header instead of the required `| File |`:

```
| Entity | Responsibility | Input→Output | Out of Scope |
```

Per `organizational_principles.rulebook.md` § Table Quality Standard 4: "Use 'File' for directories, 'Component' for architecture". This is a directory listing, so the header must be `| File |`.

**Required fix**: Replace `Entity` with `File` in the table header.

---

### Finding 10 (DESIGN) — Fallback-by-Default Pattern for a Security-Critical Configuration

**Severity**: Medium

The `ip_token_key: Option<[u8; 32]>` pattern in both `KeyFetcher::new()` and `BudgetClientConfig` allows the system to silently operate without encryption. This is a fail-open design for a security-critical feature.

The `IP_TOKEN_KEY` env var should either:
1. Be required at startup (fail-fast if absent when using server mode), or
2. If optional, the `None` arm must produce an error, not a silent passthrough

The current design requires developers to remember to set `IP_TOKEN_KEY` — a configuration gap that produces silent data corruption (ciphertext sent to OpenAI) rather than a startup failure.

---

### Finding 11 (TESTING) — E2E Tests Call Crypto Directly, Not Through Client APIs

**Severity**: Medium

`e2e_handshake_server_encrypts_client_decrypts` and similar tests verify the crypto algebra but not the system integration. The "client side" of each test creates a bare `IpTokenCrypto` instance and calls `decrypt()` directly, bypassing:
- `BudgetClient::handshake()`
- `KeyFetcher::get_key()` and `fetch_from_server()`
- The `ip_token_key` → `IpTokenCrypto` initialization path

True e2e tests would go through the actual client APIs (`BudgetClient::handshake()`, `KeyFetcher::get_key()`) and verify `get_provider_key()` returns the correct plaintext — proving the production path works. These tests are integration tests for the crypto module, not e2e tests for the full flow.

---

## Positive Observations

1. **Correct architectural decision**: Moving `IpTokenCrypto` to `iron_secrets` so both server and client can use the same implementation is the right call. The shared crypto is clean and well-structured.

2. **`iron_secrets/src/ip_token.rs` logic is sound**: The AES-256-GCM implementation (nonce generation, ciphertext format `AES256:{IV}:{ciphertext}:{tag}`, tag splitting) is correct. The `IpTokenError` variants cover all failure modes explicitly.

3. **Server-side encryption wired correctly**: Both `handshake.rs` and `agent_provider_key.rs` correctly call `state.ip_token_crypto.encrypt(&provider_key)` before returning the response. The security invariant "plaintext key never in HTTP response" holds for the server side.

4. **`IP_TOKEN_KEY` env var wired in router**: `router.rs` correctly reads `IP_TOKEN_KEY` via `parse_ip_token_key_from_env()` and passes it to both `KeyFetcher::new()` and `ProxyConfig`. The env var parsing handles missing/malformed values with appropriate tracing warnings.

5. **16 unit tests for crypto core**: `ip_token_test.rs` covers roundtrip, empty/long strings, format validation, tamper detection, wrong key, and key length validation. Good coverage of the crypto primitive.

6. **`GetProviderKeyResponse` field correctly renamed to `ip_token`**: The response struct field name matches the encrypted nature of the data — `provider_key` implied plaintext, `ip_token` correctly signals it's an encrypted token.

---

## Required Changes Before Merge

1. **Fix CRIT**: Replace `None => data.ip_token` with `None => return Err(...)` in both `key_fetcher.rs:202` and `budget_client.rs:281`
2. **Fix CRIT**: Propagate `Zeroizing<String>` without calling `.to_string()` at `key_fetcher.rs:201` and `budget_client.rs:280`
3. **Fix HIGH**: Delete `module/iron_control_api/src/ip_token.rs` (dead orphaned file)
4. **Fix HIGH**: Reformat `iron_secrets/src/ip_token.rs` to project codestyle (new-line braces, spaces in `#[ derive( ) ]` and `< >`, 2-space indent)
5. **Fix HIGH**: Add `bug_reproducer`-marked tests with 5-section docs for both fixed bugs; add 3-field fix comments in source
6. **Fix HIGH**: Add integration test for `KeyFetcher` decryption path (not direct crypto calls)
7. **Fix HIGH**: Update 3 `readme.md` files with new file entries
8. **Fix MED**: Change `Entity` → `File` in `iron_control_api/tests/readme.md` table header

**Comment by alchezar** [comment]
## Re: Finding 5 — Codestyle Violations in `iron_secrets/src/ip_token.rs`

**Status: Dismissed — superseded by PR #47**

The project codestyle was updated and agreed upon in PR #47. The new standard is `cargo fmt` with `rustfmt.toml`. The previous style (`new-line braces`, `spaces in < >`, `#[ derive( ) ]`) is no longer a requirement.

**Current enforced style:**
* Standard Rust formatting via `cargo fmt`
* Single deviation from default: **2-space indent** (configured in `rustfmt.toml` via `tab_spaces = 2`)

**In the current changes:** reformats `iron_secrets/src/crypto.rs`, `tests/crypto_test.rs`, `tests/smoke_test.rs` and related `Cargo.toml` files to comply with the new standard.

**Review by wanguardd — CHANGES_REQUESTED** [review]
## Review Summary

**Outcome**: NOT READY

### Verification Results

| Check | Status |
|-------|--------|
| Build | PASS |
| iron_secrets tests | PASS (25 tests) |
| Prior reviews | 2 (both CHANGES_REQUESTED) |
| Critical issues outstanding | 2 |
| High issues outstanding | 4 |

---

## Critical Findings (Still Present)

### Finding 1 — Silent Plaintext Fallback in `key_fetcher.rs:202`

**Severity**: CRITICAL

```rust
let api_key = match &self.ip_token_crypto {
  Some(crypto) => crypto.decrypt(&data.ip_token)...,
  None => data.ip_token,   // ← uses ciphertext as API key
};
```

When `IP_TOKEN_KEY` is absent, the encrypted `AES256:...` string is used as the API key, causing 401 errors and leaking encrypted credentials to OpenAI/Anthropic.

**Required fix**: Return error instead of silent passthrough:
```rust
None => return Err(LlmRouterError::KeyFetch(
  "IP_TOKEN_KEY not configured — cannot decrypt IP Token".into()
)),
```

---

### Finding 2 — Silent Plaintext Fallback in `budget_client.rs:281`

**Severity**: CRITICAL

Same bug in the budget handshake flow:
```rust
None => data.ip_token,   // ← ciphertext used as API key
```

**Required fix**: Same as Finding 1.

---

## High Findings

### Finding 3 — `.to_string()` Destroys `Zeroizing<String>` Memory Protection

Both `key_fetcher.rs:201` and `budget_client.rs:280` call `.to_string()` on `Zeroizing<String>`, creating unprotected copies. Propagate `Zeroizing<String>` instead.

### Finding 4 — Orphaned `iron_control_api/src/ip_token.rs`

6693-byte file exists but is never compiled (`lib.rs` only has `pub use iron_secrets::ip_token;`). Delete this dead code.

### Finding 5 — Missing Bug Fix Documentation

No `bug_reproducer(issue-NNN)` tests with 5-section format exist for the two fixed bugs.

### Finding 6 — Missing `readme.md` Updates

`ip_token_test.rs` and `ip_token_e2e_test.rs` are not listed in their directory's Responsibility Tables.

---

## Medium Findings

### Finding 7 — E2E Tests Don't Exercise Client APIs

Tests call `IpTokenCrypto::decrypt()` directly instead of through `KeyFetcher` or `BudgetClient`. Add integration tests for actual client paths.

---

## Positive Observations

1. **Sound crypto**: AES-256-GCM implementation is correct
2. **Good unit tests**: 16 tests cover edge cases thoroughly
3. **Server-side encryption works**: Both endpoints encrypt correctly
4. **Correct architecture**: Moving crypto to `iron_secrets` is right

---

## Required Changes

| Priority | Action |
|----------|--------|
| CRITICAL | Fix silent fallback in `key_fetcher.rs:202` and `budget_client.rs:281` |
| HIGH | Propagate `Zeroizing<String>` without `.to_string()` |
| HIGH | Delete orphaned `iron_control_api/src/ip_token.rs` |
| HIGH | Add bug_reproducer tests with 5-section docs |
| HIGH | Update readme.md files with new entries |
| MEDIUM | Add KeyFetcher/BudgetClient integration tests |

File: module/iron_runtime/src/llm_router/key_fetcher.rs [hunk 1]
```diff
@@ -53,31 +26,40 @@ pub struct KeyFetcher
...
+    ip_token_key: Option<[u8; 32]>,
+  ) -> Self {
     let client = Client::builder()
       .timeout(Duration::from_secs(30))
       .build()
       .expect("LOUD FAILURE: Failed to create HTTP client");
 
+    let ip_token_crypto =
+      ip_token_key.map(|key| IpTokenCrypto::new(&key).expect("LOUD FAILURE: Invalid IP Token key"));
+
```
Comment (tihilya) [inline module/iron_runtime/src/llm_router/key_fetcher.rs]: Using new type you can return error, not panic.
